providers/implementations/keymgmt/dh_kmgmt.c.in

```diff
@@ -529,162 +562,168 @@ static int dh_set_gen_seed(struct dh_gen_ctx *gctx, unsigned char *seed,
     return 1;
 }
 
-static int dh_gen_common_set_params(void *genctx, const OSSL_PARAM params[])
+struct dh_gen_set_params_st {
+    OSSL_PARAM *type;
+    OSSL_PARAM *group_name;
+    OSSL_PARAM *privlen;
+    OSSL_PARAM *pbits;
+    OSSL_PARAM *qbits;      /* HDX only */
+    OSSL_PARAM *digest;     /* HDX only */
+    OSSL_PARAM *propq;      /* HDX only */
+    OSSL_PARAM *g_index;    /* HDX only */
+    OSSL_PARAM *seed;       /* HDX only */
+    OSSL_PARAM *p_counter;  /* HDX only */
+    OSSL_PARAM *h;          /* HDX only */
+    OSSL_PARAM *generator;  /* DH only */
+};
+
+#define dhx_gen_set_params_st dh_gen_set_params_st
+
+{- produce_param_decoder('dhx_gen_set_params',
+                         (['PKEY_PARAM_FFC_TYPE',          'type',       'utf8_string'],
+                          ['PKEY_PARAM_GROUP_NAME',        'group_name', 'utf8_string'],
+                          ['PKEY_PARAM_DH_PRIV_LEN',       'privlen',    'int'],
+                          ['PKEY_PARAM_FFC_PBITS',         'pbits',      'size_t'],
+                          ['PKEY_PARAM_FFC_QBITS',         'qbits',      'size_t'],
+                          ['PKEY_PARAM_FFC_DIGEST',        'digest',     'utf8_string'],
+                          ['PKEY_PARAM_FFC_DIGEST_PROPS',  'propq',      'utf8_string'],
+                          ['PKEY_PARAM_FFC_GINDEX',        'g_index',    'int'],
+                          ['PKEY_PARAM_FFC_SEED',          'seed',       'octet_string'],
+                          ['PKEY_PARAM_FFC_PCOUNTER',      'p_counter',  'int'],
+                          ['PKEY_PARAM_FFC_H',             'h',          'int'],
+                          ['PKEY_PARAM_DH_GENERATOR',      'invalid param'],
+                         )); -}
+
+#define dh_gen_set_params_st dh_gen_set_params_st
+
+{- produce_param_decoder('dh_gen_set_params',
+                         (['PKEY_PARAM_FFC_TYPE',          'type',       'utf8_string'],
+                          ['PKEY_PARAM_GROUP_NAME',        'group_name', 'utf8_string'],
+                          ['PKEY_PARAM_DH_PRIV_LEN',       'privlen',    'int'],
+                          ['PKEY_PARAM_FFC_PBITS',         'pbits',      'size_t'],
+                          ['PKEY_PARAM_DH_GENERATOR',      'generator',  'int'],
+                          ['PKEY_PARAM_FFC_GINDEX',        'invalid param'],
```

Comment:
I am not sure automatic this 'invalid param' is worth it..
The old code at least made it clear what this was doing.
At least put a comment here if you want to persist with this approach..
i.e. There are DSA specific parameters.

Reply:
This one is tricky.  By leaving these out the parameters will be silently ignored (as parameters are meant to be).

However, the existing code detects these and raises an error to warn the user about their mistake.  That's quite the opposite of ignoring the parameters.

I went for doing what the existing code does instead of modifying things.  I assume that the current behaviour was deliberate and well reasoned.